### PR TITLE
Speedup compilepkg args handling

### PR DESCRIPTION
**Pull request by dzbarsky**
**What type of PR is this?**
Starlark cleanup

**What does this PR do? Why is it needed?**
A lot of the args between compilepkg and nogo are shared, so we can reuse the `Args` object. Cleaned up a few other things that showed up in profiles while I was here. 

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

### CODE REVIEW

**Comment by fmeum** [comment]
This is a nice improvement, including to retained memory. I considered doing this right in the nogo change, but wanted to keep it small. 

We should wait for the nogo release before we merge this.

**Comment by dzbarsky** [comment]
> This is a nice improvement, including to retained memory. I considered doing this right in the nogo change, but wanted to keep it small.
> 
> We should wait for the nogo release before we merge this.

Yep, makes sense.

**Comment by dzbarsky** [comment]
@fmeum shall we revisit this after bazel8 ships?

**Comment by fmeum** [comment]
Yes, I'll also cut the long due rules_go release!

**Comment by dzbarsky** [comment]
> Yes, I'll also cut the long due rules_go release!

OK, I've rebased it. Maybe we can try to land https://github.com/bazelbuild/rules_go/pull/4030 after the release is cut?

File: go/private/context.bzl [hunk 1]
```diff
@@ -187,14 +186,15 @@ def _builder_args(go, command = None, use_path_mapping = False):
         # Use a file rather than goroot as the latter is just a string and thus
         # not subject to path mapping.
         args.add_all("-goroot", [goroot_file], map_each = _dirname, expand_directories = False)
-    args.add("-installsuffix", installsuffix(go.mode))
-    args.add_joined("-tags", go.mode.tags, join_with = ",")
+    mode = go.mode
+    args.add("-installsuffix", installsuffix(mode))
+    if mode.tags:
+        args.add_joined("-tags", mode.tags, join_with = ",")
```
Comment (fmeum) [inline go/private/context.bzl:192]:
```suggestion
    args.add_joined("-tags", mode.tags, join_with = ",")
```

Bazel already has an early return for this case in `args.add_joined`.

**Review by fmeum — APPROVED** [review]

File: go/private/context.bzl (hunk 1 above)
Comment (dzbarsky) [reply go/private/context.bzl:192]: I saw it show up but it was only 10ms and inconsistent. I'll revert this bit

Comment (fmeum) [reply go/private/context.bzl]: I hit commit :⁠-⁠)

Comment (dzbarsky) [reply go/private/context.bzl]: Ok thanks!

**Pull request merged**